### ISSUE TITLE
Add Support for Perplexity AI

### ISSUE THREAD

**Issue by nkgokul** [logancyang/obsidian-copilot#424]
Add support for using https://www.perplexity.ai/ similar to how we can openai and azure models right now

**Comment by agu3rra** [comment]
Referencing Perplexity API docs here in case it is useful for adding support: https://docs.perplexity.ai/api-reference/chat-completions

**Comment by agu3rra** [comment]
Comparing Perplexity docs with those of [OpenAI's ChatGPT's API docs](https://platform.openai.com/docs/api-reference/making-requests), it looks like it uses the same exact endpoint `/chat/completions`. I remember the Copilot Obsidian plugin has an area for Custom API's which adhere to the specs of OpenAI's. I will make some experiments with it using my Perplexity API key and report back here.

**Comment by agu3rra** [comment]
# Tests
On a Windows 11 PC with Obsidian Copilot already enabled:
I have added one of the [available models from Perplexity](https://docs.perplexity.ai/guides/model-cards) as a custom model in the plugin settings as follows:

![custom-models-obsidian-copilot](https://github.com/user-attachments/assets/5d911026-5a05-4235-aa28-668d75be854e)

## Blocked by CORS policy
You have to ensure you mark the CORS checkbox, otherwise it will error like this:

![custom-perplexity-obsidian-copilot-blocked-cors](https://github.com/user-attachments/assets/af83c240-7931-4b82-a6c1-ab04f00e6271)

## Timeout
After enabling CORS, I see what looks like a timeout error. I tried the following and still got the same results:
1. generating a fresh API token at Perplexity (in case mine was messed up)
2. Using the base URL as https://api.perplexity.ai
3. Using the base URL as https://api.perplexity.ai/chat/completions

![obsidian-copilot-perplexity-timeout](https://github.com/user-attachments/assets/8aaa4b95-1dd2-4591-9e8e-76dd8fa70e42)

## Testing with curl
![curl-perplexity-api](https://github.com/user-attachments/assets/a3271fad-f29d-4aea-b82e-83bc59e775be)

**Comment by agu3rra** [comment]
I was unable to individually delete the custom models without resetting the plugin entirely, so I opened https://github.com/logancyang/obsidian-copilot/issues/599

**Comment by agu3rra** [comment]
@logancyang 
I re-ran the test after updating the plugin to version `2.6.2`, but the apparent timeout remains the same.
![image](https://github.com/user-attachments/assets/8d479802-1b67-49f5-8aef-e70a2cb4007b)

**Comment by logancyang** [comment]
@agu3rra Perhaps it's because perplexity api doesn't conform with openai format since it has those extra fields?

I'm considering hosting a gateway for perplexity and make it available in Copilot as a paid service, so people can subscribe with me instead of having their own perplexity API keys. WDYT? I'm a bit worried about the usage though, if I go for a subscription I may get a huge bill when people abuse it.

**Comment by agu3rra** [comment]
Hello @logancyang!

## on the error
Perhaps I can to proxy all my traffic to BURP and see if stuff coming out of Obsidian to Perplexity yields more meaningful errors. I clone the repo here to see if I can do anything useful, but I've never developed any Obsidian plugin or any TypeScript for that matter.

## on the gateway idea
I think your gateway could be a way to monetize stuff for sure. I am not sure how large of an audience Obsidian is, but there's certainly value in being able to just query AI in the context of one's Obsidian Vault. You just need to make sure any abusers are metered appropriately so you don't pay the bill on their behalf. Perhaps do it like OpenAI, where they charge you in advance and you can keep auto-charging your card to certain limits. If a person wants to abuse

I understand a gateway with added benefits is what Perplexity is if you really think about it. They make a decent UI and gear their product towards effective search. I also am a subscriber there since it functions as a gateway to other LLM's. As in, if you pay 20 USD to OpenAI you get GPT-4 usage, Claude for Anthropic on the same amount. But 20 in PPL, gives you both plus the effective search, plus easy access to open models like Llamma or the Flux.1 image generation thing.

## premium features idea?
I actually like the fact that your plugin allows one to freely select any vendor they are already using. I think it contributes to the amount of users it's receiving. So perhaps you can add some sort of premium feature that would get people to pay extra? Not even adding your gateway would be required. Right now I am strongly considering acquiring a subscription of Infranodus. I like the idea and possible insights it brings and now they added an Obsidian plugin to make it more appealing to me: https://infranodus.com/obsidian-plugin.

**Comment by logancyang** [comment]
@agu3rra These are great points. And thanks for the pointer to infranodus, I've seen one of their youtube videos once and it's quite fascinating.

I've been thinking about premium features for a while and would like to do it by providing enough extra value on top of people's API costs.

IMO my (or most) Obsidian plugin faces this monetization challenge: it's open source and completely client-side. So anyone who's a little motivated can check the code and go around any paywall-guarded logic.

Although I subscribe to a local-first and privacy-focused ideal, considering Obsidian plugins don't have an official blessed way to monetize, the only best way to offer premium features is through a hosted backend service.

I agree that the amount of users this plugin has is largely a result of the coverage of different APIs and the customizability. But it comes at a cost. "Highly customizable" means "little control" from the creator. Too many things can go wrong when a user configures in their own way. For example, a high-quality Vault QA experience requires an opinionated setup but the user is usually not an expert in that. 

I'm considering offering a "Copilot+" mode in addition to the current modes where I have a more "work out-of-the-box" setup so people can see the difference vs. their own setup.

I also thought about providing knowledge graph support like visualization and even GraphRAG, but the ROI of that is not yet convincing enough.

For now, the "premium features" I have in mind are mostly special "slash command" skills that are only available in Copilot+ mode. These skills will need to call my backend to function, but the notes and search index are still stored locally. If you are interested I can sign you up as an early tester and you can use it for free. You are welcome to suggest features too! Feel free to email me for any comments and suggestions logan@brevilabs.com

**Comment by agu3rra** [comment]
@logancyang 
Going back to the premium Copilot+ backend idea then... 

Perhaps a nitched market are Obsidian users that are not that technical and want the Copilot plugin to work out of the box? You charge them pre-paid model to avoid any upfront costs on your end and you interface to a number of different LLM vendors + host open-source models. The setup asks the user for an API key, which they acquire on your website (like Infranodus) and they call your API instead one of the vendor ones. I wonder if you can get some sort of discount from the vendors (OpenAI, Anthropic, etc) to make a good profit margin. I was watching this Lex Fridman podcast interview with Pieter Levels and that was just the part he was mentioning on monetization challenges: he would just reach out to vendors and be surprised getting 20-30% discount without a lot of effort. If you bring massive amounts of usage to their API's they want to keep you around.

I think that's a use case I'd pay for too. I am considering getting out of Perplexity Pro, now that I am paying for usage on OpenAI via the API. If Obsidian Copilot+ offered me a similar pre-paid option with the benefits of:
1. Running seamless on Obsidian.
2. Option to select from different LLM's with the click of a button.
I think it's a compelling use-case.

I wonder if there's an effective way to forecast how many users would want that. Another point in the forecast that Pieter makes is that it's really hard in his opinion convert free users into paying ones. Hence why he always launched products that asked for $$$ after a small demo/taste.

**Comment by logancyang** [comment]
PR merged. Perplexity API will work in the next release, don't forget to turn on CORS.

**Comment by the0fun** [comment]
AAA how do I connect it? even tho I'm trying to add a new model etc. I always get an error - model verification failed; model name sonar-pro. In base URL I give "https://api.perplexity.ai/chat/completions" + api key to the perplexity.

**Comment by the0fun** [comment]
<img width="656" height="662" alt="Image" src="https://github.com/user-attachments/assets/d6975ab6-1beb-4090-ba54-5ff97f55053a" /> Nevermind!!

**Comment by agu3rra** [comment]
@the0fun 
I believe Perplexity could have changed its API since this feature was first released in Obsidian-Copilot. I suggest you open a new issue, since this one was closed once the pull request was merged in 2024.
